help_function/src/onedpl_test_malloc_free.cpp

```diff
@@ -0,0 +1,226 @@
+// ====------ onedpl_test_malloc_free.cpp---------- -*- C++ -* ----===////
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//
+// ===----------------------------------------------------------------------===//
+
+#include "oneapi/dpl/algorithm"
+#include "oneapi/dpl/execution"
+#include "oneapi/dpl/iterator"
+
+#include "dpct/dpct.hpp"
+#include "dpct/dpl_utils.hpp"
+
+#include <iostream>
+
+#include <sycl/sycl.hpp>
+
+#include <sys/resource.h>
+
+template <typename String, typename _T1, typename _T2>
+int ASSERT_EQUAL(String msg, _T1 &&X, _T2 &&Y) {
+  if (X != Y) {
+    std::cout << "FAIL: " << msg << " - (" << X << "," << Y << ")" << std::endl;
+    return 1;
+  }
+  return 0;
+}
+
+int test_passed(int failing_elems, std::string test_name) {
+  if (failing_elems == 0) {
+    std::cout << "PASS: " << test_name << std::endl;
+    return 0;
+  }
+  return 1;
+}
+
+int main() {
+
+  // used to detect failures
+  int failed_tests = 0;
+  int num_failing = 0;
+
+  // Test One, test normal calls for dpct::malloc and dpct::free by allocating a
+  // certain number of bytes on device.
+  {
+    std::size_t num_elements = 5;
+    dpct::device_sys_tag device_exec;
+    dpct::tagged_pointer<dpct::device_sys_tag, void> double_ptr =
+        dpct::malloc(device_exec, num_elements * sizeof(double));
+    double num_data[] = {0.0, 1.0, 2.0, 3.0, 4.0};
+
+    dpct::get_default_queue().submit([&](sycl::handler &h) {
+      h.memcpy(double_ptr, &num_data, sizeof(double) * num_elements);
+    });
+    dpct::get_default_queue().wait();
+
+    dpct::get_default_queue().submit([&](sycl::handler &h) {
+      h.memcpy(&num_data, double_ptr, num_elements * sizeof(double));
+    });
+    dpct::get_default_queue().wait();
+
+    std::string test_name =
```

Comment:
Seems like we should check `double_ptr` values, 
Then I think we should either increment all values before copying it back, or copy it to a different host array before checking it just to confirm the copies do what we expect.

As it stands, if the copies were no-ops this would pass.

Reply:
I have added the checks of `double_ptr` and then copy to a separate buffer on the host and check this as well.